plugins/motiontrail.lua

```diff
@@ -30,43 +30,69 @@ config.plugins.motiontrail = common.merge({
   }
 }, config.plugins.motiontrail)
 
+local cc_installed = pcall(require, 'plugins.custom_caret')
+local cc_conf = config.plugins.custom_caret
 
 local function lerp(a, b, t)
+  if not (a and b and t) then return 0 end
   return a + (b - a) * t
 end
 
+local caret_index = 1
 
 local function get_caret_rect(dv)
-  local line, col = dv.doc:get_selection()
-  local x, y = dv:get_line_screen_position(line, col)
-  return x, y, style.caret_width, dv:get_line_height()
-end
+  local line, col = dv.doc:get_selection_idx(caret_index)
+  local chw = dv:get_font():get_width(dv.doc:get_char(line, col))
+   local w = style.caret_width
+  local h = dv:get_line_height()
 
+  if cc_installed then
+    local cc_shape = cc_conf.shape
+    if cc_shape ==  "block" then
+      w = chw
+    elseif cc_shape == "underline" then
+      w = chw
+      h = style.caret_width * 2
+    end
+  end
 
-local last_x, last_y, last_view
+  return w, h
+end
 
-local draw = DocView.draw
+local last_x, last_y, last_view = {}, {}, {}
 
-function DocView:draw(...)
-  draw(self, ...)
+local dv_update = DocView.update
+function DocView:update()
+  caret_index = 1
+  dv_update(self)
+end
+
+local dv_draw_caret = DocView.draw_caret
+function DocView:draw_caret(x,y)
   if not config.plugins.motiontrail.enabled or self ~= core.active_view then
+    dv_draw_caret(self, x, y)
     return
   end
 
-  local x, y, w, h = get_caret_rect(self)
+  local lsw, lsx, lsy = last_view[caret_index], last_x[caret_index], last_y[caret_index]
+  local w, h = get_caret_rect(self)
 
-  if last_view == self and (x ~= last_x or y ~= last_y) then
+  if lsw == self and (x ~= lsx or y ~= lsy) then
     local lx = x
     for i = 0, 1, 1 / config.plugins.motiontrail.steps do
-      local ix = lerp(x, last_x, i)
-      local iy = lerp(y, last_y, i)
+      local ix = lerp(x, lsx, i)
+      local iy = lerp(y, lsy, i)
+      if cc_conf.shape == "underline" then iy = iy + self:get_line_height() end
```

Comment:
You should check if the plugin is actually installed, otherwise if the shape is `underline` but `custom_caret` isn't active, the trail isn't correct.